org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/dtd/contentmodel/CMDTDDocument.java

```diff
@@ -124,6 +127,12 @@ public void startContentModel(String elementName, Augmentations augs) throws XNI
 		if (hierarchiesMap == null) {
 			hierarchiesMap = new HashMap<>();
 		}
+		if (commentsMap == null) {

```

Comment:
I think we should have just one Map (rename hierarchiesMap to elementInfoMap) which stores comment and hierarchies. To do that, I suggest that you create a private class which holds those information like:

```java
static class DTDElementInfo {

	private final Set<String> hierarchies;
	private String comment;

	public DTDElementInfo() {
		this.hierarchies = new LinkedHashSet<String>();
		this.comment = null;
	}

	public Set<String> getHierarchies() {
		return hierarchies;
	}
	
	public String getComment() {
		return comment;
	}
		
	public void setComment(String comment) {
		this.comment = comment;
	}
}
```

and elementInfoMap (rename of hierarchiesMap) should store String as key (like today) and DTDElementInfo as value.

Reply:
Hi @angelozerr , by this way, we can collect the elements' comments and their child elements' comments. But how to collect attributes' commets? We can not just put attributes and its commet to hierarchiesMap, because there are some attributes have same name.

Reply:
I forget the attribute case that you could manage like this:

	static class DTDNodeInfo {

		private String comment;

		public DTDNodeInfo() {
			this.comment = null;
		}

		public String getComment() {
			return comment;
		}

		public void setComment(String comment) {
			this.comment = comment;
		}
	}

	static class DTDElementInfo extends DTDNodeInfo {

		private final Set<String> hierarchies;
		private final Map<String, DTDNodeInfo> attributes;
		
		public DTDElementInfo() {
			this.hierarchies = new LinkedHashSet<String>();
			this.attributes = new HashMap<>();
		}

		public Set<String> getHierarchies() {
			return hierarchies;
		}
		
		public String getComment(String attrName) {
			DTDNodeInfo attr = attributes.get(attrName);
			return attr != null ? attr.getComment() : null;
		}
	}